tracer/src/Datadog.Trace/Activity/ActivityListener.cs

```diff
@@ -0,0 +1,294 @@
+            var allListenersPropertyInfo = diagnosticListenerType.GetProperty("AllListeners", BindingFlags.Public | BindingFlags.Static);
+            if (allListenersPropertyInfo is null)
+            {
+                throw new NullReferenceException("DiagnosticListener.AllListeners method cannot be found.");
+            }
+
+            var subscribeMethodInfo = allListenersPropertyInfo.PropertyType.GetMethod("Subscribe", new[] { observerDiagnosticListenerType });
+            subscribeMethodInfo?.Invoke(allListenersPropertyInfo.GetValue(null), new[] { diagnosticListenerInstance });
+        }
+
+        private static Type CreateDiagnosticObserverType(Type diagnosticListenerType, Type observerDiagnosticListenerType)
+        {
+            var assemblyName = new AssemblyName("Datadog.DiagnosticObserverListener.Dynamic");
+            assemblyName.Version = typeof(ActivityListener).Assembly.GetName().Version;
+            var assemblyBuilder = AssemblyBuilder.DefineDynamicAssembly(assemblyName, AssemblyBuilderAccess.Run);
+            var moduleBuilder = assemblyBuilder.DefineDynamicModule("MainModule");
+
+            DuckType.EnsureTypeVisibility(moduleBuilder, typeof(ActivityListener));
+
+            var typeBuilder = moduleBuilder.DefineType(
+                "DiagnosticObserver",
+                TypeAttributes.Public | TypeAttributes.Class | TypeAttributes.AutoClass | TypeAttributes.AnsiClass | TypeAttributes.BeforeFieldInit | TypeAttributes.AutoLayout | TypeAttributes.Sealed,
+                typeof(object),
+                new[] { observerDiagnosticListenerType });
+
+            var methodAttributes = MethodAttributes.Public | MethodAttributes.Virtual | MethodAttributes.Final | MethodAttributes.HideBySig;
+
+            // OnCompleted
+            var onCompletedMethod = typeBuilder.DefineMethod("OnCompleted", methodAttributes, typeof(void), Type.EmptyTypes);
+            var onCompletedMethodIl = onCompletedMethod.GetILGenerator();
+            onCompletedMethodIl.Emit(OpCodes.Ret);
+
+            // OnError
+            var onErrorMethod = typeBuilder.DefineMethod("OnError", methodAttributes, typeof(void), new[] { typeof(Exception) });
+            var onErrorMethodIl = onErrorMethod.GetILGenerator();
+            onErrorMethodIl.Emit(OpCodes.Ret);
+
+            // OnNext
+            var onSetListenerMethodInfo = typeof(DiagnosticObserverListener).GetMethod(nameof(DiagnosticObserverListener.OnSetListener), BindingFlags.Static | BindingFlags.Public)!;
+            var onNextMethod = typeBuilder.DefineMethod("OnNext", methodAttributes, typeof(void), new[] { diagnosticListenerType });
+            var onNextMethodIl = onNextMethod.GetILGenerator();
+            onNextMethodIl.Emit(OpCodes.Ldarg_1);
+            onNextMethodIl.EmitCall(OpCodes.Call, onSetListenerMethodInfo, null);
+            onNextMethodIl.Emit(OpCodes.Ret);
+
+            return typeBuilder.CreateTypeInfo()!.AsType();
```

Comment:
Shouldn't we check if `typeBuilder.CreateTypeInfo()` returns null and handle that instead of ignoring? Maybe just return `typeBuilder.CreateTypeInfo()?.AsType()` and change the return type to `Type?` and let the caller handle the null?

Reply:
you're totally right, I made the change in https://github.com/DataDog/dd-trace-dotnet/pull/2446/commits/cdcaa6275c2e0d4c77dc7473c118d5b862491310